src/transaction/log_recovery.c

```diff
@@ -1379,9 +1380,42 @@ log_rv_analysis_complete (THREAD_ENTRY * thread_p, int tran_id, LOG_LSA * log_ls
        * holding a page.
        */
       log_lsa->pageid = NULL_PAGEID;
-      log_recovery_resetlog (thread_p, &record_header_lsa, prev_lsa);
       *did_incom_recovery = true;
 
+      /*
+       * We can't use LOG_FIND_TDES (tran_index) here because the compelete log record
+       * may be the first log record seen for the transaction.
+       */
+      auto tdes = logtb_rv_find_allocate_tran_index (thread_p, tran_id, log_lsa);
```

Comment:
How about removing the `tran_index = logtb_find_tran_index (thread_p, tran_id);` in 1346L ? 
and move this line(1389L) to 1346L ? 

because, LOG_FIND_TDES () is called inside the `logtb_find_tran_index()`, and It may results in traversing the transaction table twice. 

if you do so, you also modify the `end` label which uses tran_index. (replace `tran_index` with `tdes->tran_index`)

Reply:
`logtb_rv_find_allocate_tran_index` may allocate the `tdes` object if it is the first time to call the function. In a normal case, not `is_media_crash`, if the `tdes` is already allocated, it is released to prevent it from undoing (by making it not active). If changed as you said, a `tdes` is allocated and released for nothing when the complete log record is the only record found in the log analysis phase.

Reply:
In the normal case, the complete log record is the only log record found during the analysis, `logtb_find_tran_index` returns null tran id, and just finishs. If `logtb_rv_find_allocate_tran_index` is called, it will be always allocated and released.

Reply:
Oh, I understand. Useless tdes can be allocated. Thanks for explain.